bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s

```diff
@@ -0,0 +1,851 @@
+        mov     x29, sp
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        adr     x28, 0b
+// CHECK-NOT: function f_retaasppcr
+        retaasppcr x28
+        .size f_retaasppcr, .-f_retaasppcr
+
+f_retabsppcr:
+0:
+        pacnbibsppc
+        hint    #25
+        stp     x29, x30, [sp, #-16]!
+        mov     x29, sp
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        adr     x28, 0b
+// CHECK-NOT: function f_retabsppcr
+        retabsppcr x28
+        .size f_retabsppcr, .-f_retabsppcr
+
+        .globl  f_autia171615
+        .type   f_autia171615,@function
+f_autia171615:
+        hint    #25
+        stp     x29, x30, [sp, #-16]!
+        mov     x29, sp
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        autia171615
+// CHECK-LABEL: GS-PACRET: non-protected ret found in function f_autia171615, basic block .LBB{{[0-9]+}}, at address
+// CHECK-NEXT:    The return instruction is     {{[0-9a-f]+}}:       ret
+// CHECK-NEXT:    The 1 instructions that write to the return register after any authentication are:
+// CHECK-NEXT:    1. {{[0-9a-f]+}}: ldp     x29, x30, [sp], #0x10
+// CHECK-NEXT:  This happens in the following basic block:
+// CHECK-NEXT: {{[0-9a-f]+}}:   paciasp
+// CHECK-NEXT: {{[0-9a-f]+}}:   stp     x29, x30, [sp, #-0x10]!
+// CHECK-NEXT: {{[0-9a-f]+}}:   mov     x29, sp
+// CHECK-NEXT: {{[0-9a-f]+}}:   bl      g@PLT
+// CHECK-NEXT: {{[0-9a-f]+}}:   add     x0, x0, #0x3
+// CHECK-NEXT: {{[0-9a-f]+}}:   ldp     x29, x30, [sp], #0x10
+// CHECK-NEXT: {{[0-9a-f]+}}:   autia171615
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret
+        ret
+        .size f_autia171615, .-f_autia171615
+
+        .globl  f_autib171615
+        .type   f_autib171615,@function
+f_autib171615:
+        hint    #25
+        stp     x29, x30, [sp, #-16]!
+        mov     x29, sp
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        autib171615
+// CHECK-LABEL: GS-PACRET: non-protected ret found in function f_autib171615, basic block .LBB{{[0-9]+}}, at address
+// CHECK-NEXT:    The return instruction is     {{[0-9a-f]+}}:       ret
+// CHECK-NEXT:    The 1 instructions that write to the return register after any authentication are:
+// CHECK-NEXT:    1. {{[0-9a-f]+}}: ldp     x29, x30, [sp], #0x10
+// CHECK-NEXT:  This happens in the following basic block:
+// CHECK-NEXT: {{[0-9a-f]+}}:   paciasp
+// CHECK-NEXT: {{[0-9a-f]+}}:   stp     x29, x30, [sp, #-0x10]!
+// CHECK-NEXT: {{[0-9a-f]+}}:   mov     x29, sp
+// CHECK-NEXT: {{[0-9a-f]+}}:   bl      g@PLT
+// CHECK-NEXT: {{[0-9a-f]+}}:   add     x0, x0, #0x3
+// CHECK-NEXT: {{[0-9a-f]+}}:   ldp     x29, x30, [sp], #0x10
+// CHECK-NEXT: {{[0-9a-f]+}}:   autib171615
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret
+        ret
+        .size f_autib171615, .-f_autib171615
+
+// TODO: add test to see if registers clobbered by a call are picked up.
```

Comment:
That's a good point. Potentially, any branch to an unknown (i.e. non-local) function should be assumed to clobber all the registers (caller-saved or not). I don't recall seeing that in the implementation.

Reply:
Indeed, this was not taken into account.
Now fixed in c1103cf , with associated test cases.